compiler/src/abi/json_builder.rs

```diff
@@ -0,0 +1,240 @@
+type TypeDefs<'a> = HashMap<&'a str, &'a vyp::TypeDesc<'a>>;
+
+/// TODO: Add support for events.
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Contract {
+    pub functions: Vec<Function>,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Function {
+    name: String,
+    #[serde(rename = "type")]
+    typ: FunctionType,
+    inputs: Vec<Input>,
+    outputs: Vec<Output>,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Input {
+    name: String,
+    #[serde(rename = "type")]
+    typ: VariableType,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Output {
+    name: String,
+    #[serde(rename = "type")]
+    typ: VariableType,
+}
+
+#[allow(dead_code)]
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum FunctionType {
+    Function,
+    Constructor,
+    Receive,
+    Fallback,
+}
+
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum VariableType {
+    Uint256,
+}
+
+impl ToString for VariableType {
+    fn to_string(&self) -> String {
+        match self {
+            VariableType::Uint256 => String::from("uint256"),
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum StateMutability {
+    Pure,
+    View,
+    Nonpayable,
+    Payable,
+}
+
+/// Builds a vector containing contract ABIs.
+pub fn module<'a>(module: &'a vyp::Module) -> Result<Vec<Contract>, CompileError> {
+    let type_defs: TypeDefs<'a> = module
+        .body
+        .iter()
+        .filter_map(|stmt| match stmt.node {
+            vyp::ModuleStmt::TypeDef { ref name, ref typ } => Some((name.node, &typ.node)),
+            _ => None,
+        })
+        .collect();
+
+    module
+        .body
+        .iter()
+        .filter_map(|stmt| match stmt.node {
+            vyp::ModuleStmt::ContractDef { .. } => Some(contract_def(&type_defs, &stmt.node)),
+            _ => None,
+        })
+        .collect::<Result<Vec<Contract>, CompileError>>()
+}
+
+/// Builds a contract ABI.
+pub fn contract_def<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    stmt: &'a vyp::ModuleStmt<'a>,
+) -> Result<Contract, CompileError> {
+    if let vyp::ModuleStmt::ContractDef { name: _, body } = stmt {
+        return Ok(Contract {
+            functions: body
+                .iter()
+                .map(|stmt| func_def(type_defs, &stmt.node))
+                .collect::<Result<Vec<Option<Function>>, CompileError>>()?
+                .into_iter()
+                .filter_map(|function| function)
+                .collect(),
+        });
+    }
+
+    Err(CompileError::static_str(
+        "Contract definition translation requires ContractDef parameter.",
+    ))
+}
+
+/// Builds an ABI function element, returning None if the function is private.
+pub fn func_def<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    stmt: &'a vyp::ContractStmt<'a>,
+) -> Result<Option<Function>, CompileError> {
```

Comment:
another way to keep the `Option` out of the type signature (see above for how to take the `Result` out...) is to just return the `Function` and let the caller `filter(|f| f.is_public()`

Reply:
Definitely more readable and efficient.